From 3c1a02dfb84690811217f48491ee81938036058e Mon Sep 17 00:00:00 2001
From: "Langer, Thomas" <tlanger@maxlinear.com>
Date: Fri, 16 Apr 2021 18:30:48 +0200
Subject: [PATCH] PONRTSYS-8555: Support of ethtool counters on LAN
 sub-interfaces

Add  a minimal ethtool ops for interfaces that have no physical ports.
---
 drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c b/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
index bad6c250dd0d..fa00539c3fd3 100644
--- a/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
+++ b/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
@@ -570,6 +570,7 @@ static const struct ethtool_ops ethtool_ops = {
 /* Structure of the ether tool operation in No-Phy case */
 static const struct ethtool_ops serdes_ethtool_ops = {
 	.get_drvinfo		= get_drvinfo,
+	.get_link		= ethtool_op_get_link,
 	.get_link_ksettings	= serdes_ethtool_get_link_ksettings,
 	.set_link_ksettings	= serdes_ethtool_set_link_ksettings,
 	.get_pauseparam		= get_pauseparam,
@@ -581,6 +582,14 @@ static const struct ethtool_ops serdes_ethtool_ops = {
 	.get_ethtool_stats	= get_ethtool_stats,
 };
 
+static const struct ethtool_ops subif_ethtool_ops = {
+	.get_drvinfo		= get_drvinfo,
+	.get_link		= ethtool_op_get_link,
+	.get_strings		= get_strings,
+	.get_sset_count		= get_stringset_count,
+	.get_ethtool_stats	= get_ethtool_stats,
+};
+
 static int eth_serdes_power(struct net_device *dev, bool up)
 {
 	int ret = 0;
@@ -1145,6 +1154,11 @@ static int ltq_eth_init(struct net_device *dev)
 	pr_debug("probing for number of ports = %d\n", priv->num_port);
 	pr_debug("%s called for device %s\n", __func__, dev->name);
 
+	/* First assume it is a subif, for physical ports it will be changed
+	 * in the loop later.
+	 */
+	dev->ethtool_ops = &subif_ethtool_ops;
+
 	for (i = 0; i < priv->num_port; i++) {
 		if (of_phy_is_fixed_link(priv->port[i].phy_node))
 			pr_debug("phy is fixed-link\n");
